File: DataSets/DataSet-New/LeetCodeTasks - Develop/LeetCodeTasks/3102_minimize-manhattan-distances/solutions/java/solution.java
class Solution {
    public int minMaxDist(int[][] points) {
        int n = points.length;
        int[] minX = new int[n];
        int[] maxX = new int[n];
        int[] minY = new int[n];
        int[] maxY = new int[n];

        // Initializing extreme values
        minX[0] = minY[0] = Integer.MAX_VALUE;
        maxX[0] = maxY[0] = Integer.MIN_VALUE;

        // Forward pass to compute min/max bounds excluding current from the start
        for (int i = 0; i < n; i++) {
            if (i > 0) {
                minX[i] = Math.min(minX[i - 1], points[i - 1][0] + points[i - 1][1]);
                maxX[i] = Math.max(maxX[i - 1], points[i - 1][0] + points[i - 1][1]);
                minY[i] = Math.min(minY[i - 1], points[i - 1][0] - points[i - 1][1]);
                maxY[i] = Math.max(maxY[i - 1], points[i - 1][0] - points[i - 1][1]);
            }
        }

        // Variables for backward computation of bounds
        int minXF = Integer.MAX_VALUE;
        int maxXF = Integer.MIN_VALUE;
        int minYF = Integer.MAX_VALUE;
        int maxYF = Integer.MIN_VALUE;

        int result = Integer.MAX_VALUE;

        // Backward pass to compute min/max bounds excluding current from the end
        for (int i = n - 1; i >= 0; i--) {
            if (i < n - 1) {
                minXF = Math.min(minXF, points[i + 1][0] + points[i + 1][1]);
                maxXF = Math.max(maxXF, points[i + 1][0] + points[i + 1][1]);
                minYF = Math.min(minYF, points[i + 1][0] - points[i + 1][1]);
                maxYF = Math.max(maxYF, points[i + 1][0] - points[i + 1][1]);
            }

            // Calculate the maximum Manhattan distance possible excluding the current point i
            int maxDist = 0;
            if (i > 0) {
                maxDist = Math.max(maxDist, Math.max(
                    Math.abs(maxX[i - 1] - minXF),
                    Math.abs(minX[i - 1] - maxXF)));
                maxDist = Math.max(maxDist, Math.max(
                    Math.abs(maxY[i - 1] - minYF),
                    Math.abs(minY[i - 1] - maxYF)));
            }
            if (i < n - 1) {
                maxDist = Math.max(maxDist, Math.max(
                    Math.abs(maxXF - minX[i]),
                    Math.abs(minXF - maxX[i])));
                maxDist = Math.max(maxDist, Math.max(
                    Math.abs(maxYF - minY[i]),
                    Math.abs(minYF - maxY[i])));
            }

            // Update the result with the minimum of the maximum distances found
            result = Math.min(result, maxDist);
        }

        return result;
    }
}

File: DataSets/DataSet-New/LeetCodeTasks - Develop/LeetCodeTasks/3102_minimize-manhattan-distances/solutions/python3/solution.py
class Solution:
    def minimumDistance(self, points: list[list[int]]) -> int:
        def manhattan(p1, p2):
            return abs(p1[0] - p2[0]) + abs(p1[1] - p2[1])
        
        def get_extreme_points():
            max_x, max_y = -float('inf'), -float('inf')
            min_x, min_y = float('inf'), float('inf')
            for x, y in points:
                if x + y > max_x + max_y:
                    max_x, max_y = x, y
                if x + y < min_x + min_y:
                    min_x, min_y = x, y
                if x - y > max_x - max_y:
                    max_x_neg, max_y_neg = x, y
                if x - y < min_x_neg - min_y_neg:
                    min_x_neg, min_y_neg = x, y
            return (max_x, max_y), (min_x, min_y), (max_x_neg, max_y_neg), (min_x_neg, min_y_neg)
        
        n = len(points)
        if n == 3:  # with only 3 points, removing any one will leave us with only one pair to consider
            return min(manhattan(points[0], points[1]), manhattan(points[0], points[2]), manhattan(points[1], points[2]))
        
        # Finding the extreme points based on `x + y` and `x - y` criteria
        max1, min1, max2, min2 = get_extreme_points()
        
        # Check the max distances by removing one point at a time
        min_max_distance = float('inf')
        for i in range(n):
            current_point = points[i]
            candidates = [max1, min1, max2, min2]
            # Consider other extreme points when current point is removed
            max_distance_when_removed = 0
            for p1 in candidates:
                if p1 == current_point:
                    continue
                for p2 in candidates:
                    if p2 == current_point or p1 == p2:
                        continue
                    max_distance_when_removed = max(max_distance_when_removed, manhattan(p1, p2))
            min_max_distance = min(min_max_distance, max_distance_when_removed)
        
        return min_max_distance
